LVGL: Documentation of the LVGL performance. (#146)
espressif/esp-bsp: components/esp_lvgl_port/docs/performance.md

# LCD & LVGL Performance

This document provides steps, how to set up your LCD and LVGL port for the best performance and comparing the difference settings. All settings and measurements are valid for Espressif's chips.

## Process of the performance measuring

There are two main processes for measurement performance on LCD:

1. Measure time, how long it takes to refresh the whole screen.

> 2. Use `lv_demo_benchmark()` to measure FPS.

Reviewer: We can also add this: 3. Use `lv_demo_music()` (with autoplay) to measure FPS.  
Because the result of benchmark demo is weighted FPS (more like score), and the result of music demo is average FPS. Many users may care about actual FPS.

Reviewer: And the configuration `LV_DISP_DEF_REFR_PERIOD` affects the results of performance testing very much. Though reducing the value of this param won't improve actual LVGL performance normally, it will increase weighted or average FPS. We should mention it.

Author: Thank you for info. I will add it into readme and add it into my example table too

Author: Updated documentation.

# LCD & LVGL Performance

> This document provides steps, how to set up your LCD and LVGL port for the best performance and comparing the difference settings. All settings and measurements are valid for Espressif's chips.

Reviewer 2: ```suggestion
This document provides steps, how to set up your LCD and LVGL port for the best performance and comparison of different settings. All settings and measurements are valid for Espressif's chips.
```

## Process of the performance measuring

> There are three main processes for measurement performance on LCD:

Reviewer 2: Maybe: In this document we will use following metrics for performance evaluation:

> 1. Measure time, how long it takes to refresh the whole screen.

Reviewer 2: ```suggestion
1. Measure time needed for refreshing the whole screen.
```

> 2. Use `lv_demo_benchmark()` to measure FPS (weighted FPS).
> 3. Use `lv_demo_music()` to measure FPS (average FPS).

Reviewer 2: ```suggestion
2. Use LVGL's `lv_demo_benchmark()` -test suite- to measure FPS (weighted FPS).
3. Use LVGL's `lv_demo_music()` -demo application- to measure FPS (average FPS).
```

## The settings on ESP32 chips which have impact for LCD and LVGL performance

These options and settings have impact for LCD performance (FPS). Some options have only small difference of FPS (e.g. 1 FPS), and some of them are better. Usually it depends on application, resources and size of screen. Also the HW interface and clock speed is really important (not included in this sheet).

### LVGL Buffer size

* The size of **LVGL buffer** and **double buffering enables** has big impact on performance. 

> * The best is use the same size of the buffer, like size of the LCD in pixels and enable double buffering. 

Reviewer 2: This is pretty far from optimal recommendation, from our tests it seems that having buffer larger than 25% of the screen does not bring almost any performance boost

> * If there is not enough RAM inside the CPU, you can use PSRAM (DMA is not available).

Reviewer 2: This is one of the most important parts, could we elaborate a little more?

### Compiler optimization level

> Recommended level is "Performance" for good results. The "Performance" setting causes the compiled code to be larger and faster, but will be easier to correlate code addresses to source file lines.

Reviewer 2: 'easier to correlate code addresses to source file lines' isn't it the other way around?

* `CONFIG_COMPILER_OPTIMIZATION_PERF=y`

### CPU frequency

The CPU frequency has a big impact on LCD performance. The recommended value is 240 MHz.

* `CONFIG_ESP_DEFAULT_CPU_FREQ_MHZ_240=y`

### Flash frequency

> The flash clock frequency has a big impact on LCD performance. The recommended value is 120 MHz. When PSRAM is used, the PSRAM clock frequency must be the same as a flash clock frequency.

Reviewer 2: Not all flash chips support 120MHz frequency.

This is not true either ' When PSRAM is used, the PSRAM clock frequency must be the same as a flash clock frequency.' 
Checkout this docs: https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-guides/flash_psram_config.html

Author: I don't know too much about it, but when I want to use PSRAM with LCD, I had to set same frequencies - otherwise the screen image was broken.

* `CONFIG_ESPTOOLPY_FLASHFREQ_120M=y`

### Flash mode

The recommended flash mode is QIO, which uses two additional GPIOs for SPI flash communication.

> * `CONFIG_ESPTOOLPY_FLASHMODE_QIO=y`

Reviewer 2: This will depend on the HW configruation that user has. We can only recommend using quad flash for graphics applications

Author: All performance settings are only recommendations

### Fast LVGL memory

This option puts some LVGL functions into IRAM for speed up.

* `CONFIG_LV_ATTRIBUTE_FAST_MEM_USE_IRAM=y`

### Affinity main task to second core

> The main LVGL task can be processed on the second core of the CPU. It can increase performance.

Reviewer 2: We should mention here that this is possible only on dual-core SoCs :D

### Compiler optimization level

Recommended level is "Performance" for good results. The "Performance" setting causes the compiled code to be larger and faster.

* `CONFIG_COMPILER_OPTIMIZATION_PERF=y`

### CPU frequency

The CPU frequency has a big impact on LCD performance. The recommended value is maximum -> 240 MHz.

* `CONFIG_ESP_DEFAULT_CPU_FREQ_MHZ_240=y`

### Flash frequency and mode

The flash clock frequency and mode (bus width) has a big impact on LCD performance. Your choices will be limited by your hardware flash configuration. The higher the frequency and bus width, the better.

* `CONFIG_ESPTOOLPY_FLASHFREQ_120M=y`
* `CONFIG_ESPTOOLPY_FLASHMODE_QIO=y` or `CONFIG_ESPTOOLPY_OCT_FLASH` on supported chips

More information about SPI Flash configuration can be found in [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-guides/flash_psram_config.html).

### Fast LVGL memory

This option puts the most frequently used LVGL functions into IRAM for execution speed up.

* `CONFIG_LV_ATTRIBUTE_FAST_MEM_USE_IRAM=y`

### Affinity main task to second core

> //@todo it is true that it is better if the LVGL task and its associated interrupts are handled by CPU1, but this is not the correct way of achieving that. esp_lvgl_port should be extended to allow this

Reviewer 2: @espzav sorry, I left this todo in my commit
